### ISSUE TITLE
Nasty Exception with Scope/Provider

### ISSUE THREAD

**Issue by zpqrtbnk** [umbraco/Umbraco-CMS#5151]
As reported in #5121 some nasty exceptions were found in logs:

~~~~
System.ArgumentException: An item with the same key has already been added.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at Umbraco.Core.Scoping.ScopeProvider.SetCallContextObject(String key, IInstanceIdentifiable value)
   at Umbraco.Core.Scoping.ScopeProvider.SetAmbient(Scope scope, ScopeContext context)
   at Umbraco.Core.Scoping.ScopeProvider.CreateScope(IsolationLevel isolationLevel, RepositoryCacheMode repositoryCacheMode, IEventDispatcher eventDispatcher, Nullable`1 scopeFileSystems, Boolean callContext, Boolean autoComplete)
   at Umbraco.Core.Services.Implement.UserService.GetUserById(Int32 id)
   at Umbraco.Core.Services.Implement.UserService.GetProfileById(Int32 id)
   at Umbraco.Core.ContentExtensions.GetCreatorProfile(IContentBase content, IUserService userService)
   at Umbraco.Examine.ContentValueSetBuilder.<GetValueSets>d__3.MoveNext()
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at Examine.LuceneEngine.Providers.LuceneIndex.ForceProcessQueueItems(Boolean block) in C:\projects\examine\src\Examine\LuceneEngine\Providers\LuceneIndex.cs:line 879
~~~~

~~~~
System.NullReferenceException: Object reference not set to an instance of an object.
   at Umbraco.Core.Scoping.Scope.<>c__DisplayClass72_0.<RobustExit>b__2()
   at Umbraco.Core.Scoping.Scope.TryFinally(Int32 index, Action[] actions)
   at Umbraco.Core.Scoping.Scope.TryFinally(Int32 index, Action[] actions)
   at Umbraco.Core.Scoping.Scope.TryFinally(Int32 index, Action[] actions)
   at Umbraco.Core.Scoping.Scope.RobustExit(Boolean completed, Boolean onException)
   at Umbraco.Core.Scoping.Scope.DisposeLastScope()
   at Umbraco.Core.Scoping.Scope.Dispose()
   at Umbraco.Core.Services.Implement.UserService.GetUserById(Int32 id)
   at Umbraco.Core.Services.Implement.UserService.GetProfileById(Int32 id)
   at Umbraco.Core.ContentExtensions.GetCreatorProfile(IContentBase content, IUserService userService)
   at Umbraco.Examine.ContentValueSetBuilder.<GetValueSets>d__3.MoveNext()
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at Examine.LuceneEngine.Providers.LuceneIndex.ForceProcessQueueItems(Boolean block) in C:\projects\examine\src\Examine\LuceneEngine\Providers\LuceneIndex.cs:line 879
~~~~

?

**Comment by zpqrtbnk** [comment]
Both exception happen from `LuceneIndex.ForceProcessQueueItems` which seems to only happens when the index is disposed - and is that when the site shuts down?

**Comment by zpqrtbnk** [comment]
Also reported by Shan today
~~~
Exception
System.ArgumentException: An item with the same key has already been added.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at Umbraco.Core.Scoping.ScopeProvider.SetCallContextObject(String key, IInstanceIdentifiable value)
   at Umbraco.Core.Scoping.ScopeProvider.CreateScope(IsolationLevel isolationLevel, RepositoryCacheMode repositoryCacheMode, IEventDispatcher eventDispatcher, Nullable`1 scopeFileSystems, Boolean callContext, Boolean autoComplete)
   at Umbraco.Core.Services.Implement.UserService.GetUserById(Int32 id)
   at Umbraco.Core.Services.Implement.UserService.GetProfileById(Int32 id)
   at Umbraco.Examine.ContentValueSetBuilder.<GetValueSets>d__3.MoveNext()
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at Examine.LuceneEngine.Providers.LuceneIndex.ForceProcessQueueItems(Boolean block)
~~~

**Comment by filefontaine** [comment]
Hi, this report is a bit old but i have the same error with 8.1.0 and 8.1.1 : 
´´´
System.NullReferenceException: Object reference not set to an instance of an object.

   at Umbraco.Core.Scoping.Scope.<>c__DisplayClass72_0.<RobustExit>b__2() in d:\a\1\s\src\Umbraco.Core\Scoping\Scope.cs:line 439

   at Umbraco.Core.Scoping.Scope.TryFinally(Int32 index, Action[] actions) in d:\a\1\s\src\Umbraco.Core\Scoping\Scope.cs:line 474

   at Umbraco.Core.Scoping.Scope.TryFinally(Int32 index, Action[] actions) in d:\a\1\s\src\Umbraco.Core\Scoping\Scope.cs:line 478

   at Umbraco.Core.Scoping.Scope.TryFinally(Int32 index, Action[] actions) in d:\a\1\s\src\Umbraco.Core\Scoping\Scope.cs:line 478

   at Umbraco.Core.Scoping.Scope.RobustExit(Boolean completed, Boolean onException) in d:\a\1\s\src\Umbraco.Core\Scoping\Scope.cs:line 422

   at Umbraco.Core.Scoping.Scope.Dispose() in d:\a\1\s\src\Umbraco.Core\Scoping\Scope.cs:line 365

   at Umbraco.Core.Services.Implement.UserService.GetUserById(Int32 id) in d:\a\1\s\src\Umbraco.Core\Services\Implement\UserService.cs:line 718

   at Umbraco.Core.Services.Implement.UserService.GetProfileById(Int32 id) in d:\a\1\s\src\Umbraco.Core\Services\Implement\UserService.cs:line 674

   at Umbraco.Examine.ContentValueSetBuilder.<GetValueSets>d__3.MoveNext() in d:\a\1\s\src\Umbraco.Examine\ContentValueSetBuilder.cs:line 43

   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()

   at Examine.LuceneEngine.Providers.LuceneIndex.ForceProcessQueueItems(Boolean block) in C:\projects\examine\src\Examine\LuceneEngine\Providers\LuceneIndex.cs:line 879
´´´

**Comment by Shazwazza** [comment]
@zpqrtbnk this is a super old one but had never been looked into and it's reared its head up again. 

I'm pretty sure i know the underlying problem and have created a PR here for it https://github.com/umbraco/Umbraco-CMS/pull/7994

Any chance you can glance over it and confirm my suspicions ?

**Comment by Shazwazza** [comment]
Fixed in #7994
